### PR TITLE
Getting multi-node (hosts) config working

### PR DESCRIPTION
**Pull request by tractorcow**
I'm working on a multi-server deployment; In order to make my life easier I've hacked beam to support `hosts` to specify multiple destination servers (whereas you can still use `host` to specify a single server).

At the moment it's very tentative, and doesn't synchronise very well, and only works for rsync.

### CODE REVIEW

**Comment by tractorcow** [comment]
Some more updates; I've implemented a result-merge for multi-nodes and it seems to work well enough for rsync.

I've done a bit of phpdoc / type cleanup, but I'll need to do some more later.

**Comment by tractorcow** [comment]
Requesting initial peer review from @stevie-mayhew before I add in ftp support and fix the tests. :)

**Comment by stevie-mayhew** [comment]
@tractorcow go for it, LGTM

@camspiers or @stecman do either of you have opinions on this approach?

**Comment by tractorcow** [comment]
Patched unit tests (run locally, at least).

**Comment by stecman** [comment]
Interesting, @tractorcow. Are you planning to identify the servers in `combineResults`? That's the only concern I have really.

Something like this would work in simple cases:

```
[all] [sent] [checksum,permissions] vendor/autoload.php
[server1] [deleted] [missing] some/rando/file.txt
[server1, server3] [deleted] [missing] some/file/on/two/servers.txt
```
I guess an alternative approach to this would be making a slightly better public API in Beam so you could use it as a package and implement the multi-server part in your own script:

```php
// php -f multibeam.php s1 s2 s3 s4 s5
foreach ($servers as $serverName) {
    $beam = new Beam('up', $serverName);
    $result = $beam->getChanges();
    // ...
}
```

**Comment by tractorcow** [comment]
Well, I considered that, but there isn't really a need to consider all permutations (i.e. no file needs to be added on one server, but deleted on another) so there did seem to be a distinct hierarchy of change severity.

Thus I considered that it's probably enough to simply display the MOST severe change for any server (e.g. if modified on one server, but added to another, just treat it as an "add").

However, what I could do is split the change action (most severe) into the display component (could display all distinct states across all servers).

E.g. simplified version of above; Action is displayed in first column, list of all distinct states in second column. Instead of listing each change per-server, simply summarise the list of all servers with changes in the bottom.

```
[sent] [checksum,permissions] vendor/autoload.php
[deleted] [missing] some/rando/file.txt
[deleted] [missing] some/file/on/two/servers.txt
Changes detected on server [server1, server2]
```

I could adjust this to display server changes per-line, but probably information overload.

```
[sent] [server1=checksum,server2=permissions] vendor/autoload.php
[deleted] [missing] some/rando/file.txt
[deleted] [missing] some/file/on/two/servers.txt
```

**Comment by tractorcow** [comment]
> I guess an alternative approach to this would be making a slightly better public API in Beam so you could use it as a package and implement the multi-server part in your own script:


I could turn `Server` into an interface, and abstract into a `MultiNodeServer` and `SingleServer`. :)

**Comment by stecman** [comment]
> ie. no file needs to be added on one server, but deleted on another

If you assume all the configuration is correct and the servers are perfectly in sync, sure. The kinds of cases I'm thinking of are:

- Something different was deployed to one of the servers
- Someone manually messed with one of the servers
- One of the servers in the list is incorrect (completely different content that expected)

> Thus I considered that it's probably enough to simply display the MOST severe change for any server (e.g. if modified on one server, but added to another, just treat it as an "add").

Makes sense. As a user though, I'd still want a way to figure out which server any unexpected results are coming from so I could go check it out manually.

Your suggestion of grouping different results seems sane, though I guess keep in mind we made an effort to stream output from rsync as it was previously painful to wait for the dry-run to complete before seeing anything results in the console. Maybe you could still stream the full output from all servers, but summarise/warn any differences at the end:

```
+---------+------+----------+---------+---------+------------+
| Server  | Sent | Received | Created | Deleted | Attributes |
+---------+------+----------+---------+---------+------------+
| server1 |    3 |          |       1 |       1 |          1 |
| server2 |    3 |          |       1 |    4368 |          1 |
| server3 |    5 |          |       5 |     199 |            |
+---------+------+----------+---------+---------+------------+
```

>  I could turn Server into an interface

Related: [The Great Refactoring](https://github.com/heyday/beam/wiki/Road-map).

**Comment by tractorcow** [comment]
Ah sounds fun. That being said, I'm deploying my year-long's effort to production now and I'll probably not have time to do any significant refactor until after that's complete.

Adjusting / tweaking console output is probably on the table in the short term.

**Comment by tractorcow** [comment]
It looks like I've mad an incorrect assumption; Beam is actually outputting the rsync dry run output of each server, and emitting it to the console, which is then getting parsed in before we merge the outputs.

By the time we've merged the output, we have already send that to the standard out.

**Comment by tractorcow** [comment]
image

**Comment by tractorcow** [comment]
How does this look? Enough to get it over the line? :P

**Comment by tractorcow** [comment]
I've added a server change count to each line in magenta.

E.g. 3/3 means all servers will have this change. 1/3 means only 1.

image

**Comment by stevie-mayhew** [comment]
Thanks @tractorcow

**Comment by tractorcow** [comment]
Thanks @stevie-mayhew for the merge. :)

File: src/DeploymentProvider/Rsync.php [hunk 1]
```diff
@@ -99,44 +102,84 @@ public function configure(InputInterface $input, OutputInterface $output)
...
+        foreach ($this->getTargetPaths() as $server => $targetPath) {
+            $result = $this->deploy(
+                $this->buildCommand(
+                    $this->beam->getLocalPath(),
+                    $targetPath,
+                    $dryrun
+                ),
+                $output,
+                true // silence per-server output
+            );
```
Comment (stecman) [inline src/DeploymentProvider/Rsync.php]: Can single server deployments please keep the original behaviour of streaming output directly? This re-introduces a delay that we specifically got rid of.

(It's fine to keep the delay for multi-server deployments for now though)

**Comment by stecman** [comment]
Also, any plan to make the rsync commands run in parallel, @tractorcow?

**Comment by stecman** [comment]
Not sure how I feel about removing `composer.lock` by the way - it means we're now relying on all of our dependencies to follow semver perfectly.. What was the reasoning behind removing it?

File: src/DeploymentProvider/Rsync.php (hunk 1 above)
Comment (tractorcow) [reply src/DeploymentProvider/Rsync.php]: Good idea actually; I can make it conditional.

**Comment by tractorcow** [comment]
@stecman lock file would be ignored when installed as a local or global dependency, meaning the tests would only run against a legacy set of dependencies (which might pass), but then break when installed. You'd run the risk of a false positive (or false negative) based on some outdated historic state.

**Comment by tractorcow** [comment]
> Also, any plan to make the rsync commands run in parallel, @tractorcow?

That'd be the excuse I need to rewrite it in python. :)

**Comment by stecman** [comment]
> the tests would only run against a legacy set of dependencies (which might pass), but then break when installed. You'd run the risk of a false positive (or false negative) based on some outdated historic state.

Ah right. Beam has historically not been used as a library, so I was only thinking of the "clone -> composer install -> execute/compile-to-phar" scenario. Could have both - the tests for the shell completion library [run against the newest and oldest dependency configurations](https://github.com/stecman/symfony-console-completion/blob/master/.travis.yml) which could be applied here.

> Parallel ... rewrite in Python :smile: 

All of the machinery to do this is actually in Beam already - just use `Process::start($outputHandler)` instead of the blocking `Process::run()`, then loop calling `Process::updateStatus` to feed the output handlers until all of the processes are finished.

**Comment by tractorcow** [comment]
I'll give it a look when I get some free time. :)

File: src/DeploymentProvider/Rsync.php (hunk 1 above)
Comment (tractorcow) [reply src/DeploymentProvider/Rsync.php:129]: Sorry it's been a year https://github.com/heyday/beam/pull/63

**Comment by tractorcow** [comment]
https://github.com/heyday/beam/pull/63

**Pull request merged**